src/lib.rs

```diff
@@ -7,31 +7,82 @@
 
 #![no_std]
 #![cfg_attr(feature = "unstable", feature(core_intrinsics))]
-#![cfg_attr(feature = "unstable", feature(const_generics))]
 #![cfg_attr(feature = "unstable", feature(slice_range))]
-#![cfg_attr(feature = "unstable", allow(incomplete_features))]
+#![cfg_attr(feature = "unstable", feature(slice_ptr_get))]
+#![cfg_attr(feature = "unstable", feature(slice_ptr_len))]
+#![cfg_attr(feature = "very_unstable", feature(const_slice_ptr_len))]
+#![cfg_attr(feature = "very_unstable", feature(const_panic))]
+#![cfg_attr(feature = "very_unstable", feature(const_fn_trait_bound))]
+#![cfg_attr(feature = "very_unstable", feature(const_fn_fn_ptr_basics))]
+#![cfg_attr(feature = "very_unstable", feature(const_trait_impl))]
+#![cfg_attr(feature = "very_unstable", feature(const_mut_refs))]
+#![cfg_attr(feature = "very_unstable", allow(incomplete_features))]
 #![cfg_attr(all(feature = "unstable", test), feature(slice_as_chunks))]
 #![warn(missing_docs)]
+#![deny(unsafe_op_in_unsafe_fn)]
 
-use access::{ReadOnly, ReadWrite, Readable, Writable, WriteOnly};
+use access::{Access, ReadOnly, ReadWrite, WriteOnly};
 use core::{
     fmt,
     marker::PhantomData,
-    ops::Deref,
-    ops::{DerefMut, Index, IndexMut},
-    ptr,
-    slice::SliceIndex,
+    mem,
+    ptr::{self, NonNull},
 };
 #[cfg(feature = "unstable")]
 use core::{
     intrinsics,
     ops::{Range, RangeBounds},
-    slice::range,
+    slice::{range, SliceIndex},
 };
 
 /// Allows creating read-only and write-only `Volatile` values.
 pub mod access;
 
+/// TODO
+///
+/// ## Examples
+///
+/// Accessing a struct field:
+///
+/// ```
+/// # extern crate core;
+/// use volatile::{VolatilePtr, map_field};
+/// use core::ptr::NonNull;
+///
+/// struct Example { field_1: u32, field_2: u8, }
+/// let mut value = Example { field_1: 15, field_2: 255 };
+/// let mut volatile = unsafe { VolatilePtr::new_read_write(NonNull::from(&mut value)) };
+///
+/// // construct a volatile reference to a field
+/// let field_2 = map_field!(volatile.field_2);
+/// assert_eq!(field_2.read(), 255);
+/// ```
+#[macro_export]
+macro_rules! map_field {
+    ($volatile:ident.$place:ident) => {
+        unsafe {
+            $volatile.map(|ptr| {
+                core::ptr::NonNull::new(core::ptr::addr_of_mut!((*ptr.as_ptr()).$place)).unwrap()
+            })
+        }
+    };
+}
```

Comment:
This allows creating `VolatilePtr`s to unaligned fields within packed structs, but `core::ptr::read_volatile` requires pointers to be properly aligned.

Reply:
Thanks, should be fixed by #28.